### PR TITLE
Site editor sidebar: consolidate rename component and use TemplateActions component for patterns single view 

### PR DESCRIPTION
**Pull request by ramonjd**
## What's happening here?

This PR:

- unifies the export/duplicate/rename/delete actions dropdown into a single actions component 
- renames 'Template Actions' to 'Pattern actions' in anticipation of the [great unification](https://github.com/WordPress/gutenberg/pull/54271#issuecomment-1760612399).

https://github.com/WordPress/gutenberg/assets/6458278/66221a87-82b1-42fb-924b-fb8c02a31aa6

This PR builds on the work in:

- https://github.com/WordPress/gutenberg/pull/54173#pullrequestreview-1610241523

## TODO

- [ ] Write some e2e tests

## Why? 

#54173 added rename/delete actions to Template parts. Why should patterns miss out?

## How?

- Rejigging `<TemplateActions />` (now renamed to `<PatternActions />`) to handle Pattern post types
- Consolidating template and pattern rename functionality into one `<RenameMenuItem />` component
- Consolidating template and pattern duplicate functionality into one `<DuplicateMenuItem />` component

Note: I haven't looked at the new dataview yet. It might require some refactoring if it exits experimental:
Screenshot 2023-12-04 at 4 22 50 pm

## Testing

1. Create a custom pattern in the Site Editor
2. From the sidebar, navigate to Patterns > My Patterns > Your new custom pattern. Click on your pattern's title so that the single pattern view appears in the side bar.
3. Ensure that the rename, duplicate, clear, export and delete functionality works
4. Do the same for patterns on the main patterns page, `/wp-admin/site-editor.php?path=%2Fpatterns`
5. Test out custom templates and template parts to ensure there are no regressions
6. Check that single pattern renaming et.al from the command center still works as per https://github.com/WordPress/gutenberg/pull/55188
7. Check that the actions work on "Manage all" pages for template and template parts, e.g., `/wp-admin/site-editor.php?path=%2Fwp_template%2Fall` and `/wp-admin/site-editor.php?path=%2Fwp_template_part%2Fall`

### CODE REVIEW

File: packages/edit-site/src/utils/constants.js [hunk 1]
```diff
@@ -1,5 +1,42 @@
+/**
```
Comment (ramonjd) [inline packages/edit-site/src/utils/constants.js]: Moving all this work over to:

- https://github.com/WordPress/gutenberg/pull/54484

File: packages/edit-site/src/components/sidebar-navigation-screen-template/index.js [hunk 2]
```diff
@@ -98,6 +98,11 @@ export default function SidebarNavigationScreenTemplate() {
 		postId
 	);
 
+	// The absence of a post type in the query params for templates
+	// indicates the user has arrived at the template via the "manage all"
+	// page and the back button should return them to that list page.
+	const backPath =
+		postType !== 'wp_template' ? `/${ postType }/all` : `/${ postType }`;
```
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-navigation-screen-template/index.js]: Ensure that we go back to the right screen after deleting a custom template or template part:
1. When we're arriving from the site editor sidebar, then back to that. 
2. If we're arriving from the manage all screen, stay on that (`/all`)

File: packages/edit-site/src/components/template-actions/index.js [hunk 3]
```diff
@@ -22,7 +23,12 @@ import { store as editSiteStore } from '../../store';
...
 import RenameMenuItem from './rename-menu-item';
-import { TEMPLATE_POST_TYPE } from '../../utils/constants';
+import {
+	TEMPLATE_POST_TYPE,
+	TEMPLATE_PART_POST_TYPE,
+	POST_TYPE_LABELS,
+	PATTERN_TYPES,
+} from '../../utils/constants';
 
 export default function TemplateActions( {
```
Comment (ramonjd) [inline packages/edit-site/src/components/template-actions/index.js]: Does the name of this component still make sense?

Comment (aaronrobertshaw) [reply packages/edit-site/src/components/template-actions/index.js]: Not really to me. 

My understanding is long term we'll be moving to everything being "patterns", which includes template parts and templates(?).  I'd lean towards templates being specialised patterns rather than the other way around.

File: packages/edit-site/src/components/template-actions/rename-menu-item.js [hunk 4]
```diff
@@ -53,29 +70,33 @@ export default function RenameMenuItem( { template, onClose } ) {
...
-				template.type === TEMPLATE_POST_TYPE
-					? __( 'Template renamed.' )
-					: __( 'Template part renamed.' ),
+				sprintf(
+					// translators: %1$s is a post type label, e.g., Template, Template Part or Pattern. %2$s is the new title.
+					__( '%1$s renamed to "%2$s".' ),
+					POST_TYPE_LABELS[ postType ] ??
+						POST_TYPE_LABELS.wp_template,
+					editedTitle
+				),
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/template-actions/rename-menu-item.js]: Nit: I think we were trying to avoid ending the snackbar notices with a quoted name as the punctuation should differ between languages and even between British and American English. Perhaps it is best to leave this as "post type renamed"?

File: packages/edit-site/src/components/template-actions/rename-menu-item.js [hunk 5]
```diff
@@ -53,29 +70,33 @@ export default function RenameMenuItem( { template, onClose } ) {
...
-			const fallbackErrorMessage =
-				template.type === TEMPLATE_POST_TYPE
-					? __( 'An error occurred while renaming the template.' )
-					: __(
-							'An error occurred while renaming the template part.'
-					  );
+			const fallbackErrorMessage = sprintf(
+				// translators: %s is a post type label, e.g., Template, Template Part or Pattern.
+				__( 'An error occurred while renaming the %s.' ),
+				POST_TYPE_LABELS[ postType ] ?? POST_TYPE_LABELS.wp_template
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/template-actions/rename-menu-item.js]: This code is used in multiple places would it aid readability to define it once in a meaningful variable e.g. `postTypeLabel`

**Review by aaronrobertshaw — APPROVED** [review]
Appreciate the continued iterations here @ramonjd, especially when they clean up the code some ✨ 

✅ User-created patterns show the actions menu on their view page and it works as expected
✅ Theme patterns can't be edited directly, so aren't impacted
✅ Unedited template parts do not show an actions menu
✅ Customized template parts show an actions menu containing only "clear customizations"
✅ User-created template parts show menu and can be renamed or deleted successfully
✅ After deleting a pattern or template part I was returned to the expected page

Looks good. I left a couple of minor comments regarding possible tweaks but I'll leave those to your discretion.

Also, while testing this, it kind of stuck out that we don't have the "duplicate" option under the actions menu. What do you think of that as a further follow-up?

**Comment by glendaviesnz** [comment]
I wonder if we should hold off merging this until the 6.4 release is finalised, just so we don't diverge too much from the 16.7 RC branch if any last-minute bug fixes are needed?

**Comment by ramonjd** [comment]
> I wonder if we should hold off merging this until the 6.4 release is finalised, just so we don't diverge too much from the 16.7 RC branch if any last-minute bug fixes are needed?

👍🏻 

I'll lean on your experience. Happy to hold off and keep it rebased until we're ready. 

It's low priority I'd venture and mainly janitorial, though it does bring the rename/delete experience to the patterns single view, which is lacking at the moment.

File: packages/edit-site/src/components/template-actions/index.js (hunk 3 above)
Comment (ramonjd) [reply packages/edit-site/src/components/template-actions/index.js]: Thanks for the confirmation. 

I think we could rename it to `pattern-actions` or something. I'll make a note to do this as a follow up 👍🏻

**Comment by ramonjd** [comment]
Thanks for testing this one @aaronrobertshaw 

> Also, while testing this, it kind of stuck out that we don't have the "duplicate" option under the actions menu. What do you think of that as a further follow-up?

Yes! While I was tooling around with this PR I did dive into `packages/edit-site/src/components/page-patterns/grid-item.js` to see what we could abstract. I think it's doable and a good candidate for a follow up PR. Cheers!


### Follow ups

- [ ] Rename `template-actions` directory and its components to align more neatly with patterns. https://github.com/WordPress/gutenberg/pull/54271#pullrequestreview-1641315510
- [ ] Add "Duplicate" to the actions menu for custom patterns

**Comment by aaronrobertshaw** [comment]
>  I think it's doable and a good candidate for a follow up PR.

FYI there are also some potential changes to the duplication of patterns that it may help to be aware of: https://github.com/WordPress/gutenberg/pull/54764

**Comment by ramonjd** [comment]
> I wonder if we should hold off merging this until the 6.4 release is finalised, just so we don't diverge too much from the 16.7 RC branch if any last-minute bug fixes are needed?

@glendaviesnz Do you think we should still hold off? No rush - I've rebased and tested against the command center renaming changes in https://github.com/WordPress/gutenberg/pull/55188.

**Comment by aaronrobertshaw** [comment]
> Do you think we should still hold off? No rush - I've rebased and tested against the command center renaming changes in https://github.com/WordPress/gutenberg/pull/55188.

There's also some refactoring to avoid duplicating code in https://github.com/WordPress/gutenberg/pull/55292 which adds the duplicate pattern command. Further work is in the pipeline to add commands for renaming/duplicating template parts.

It makes sense to me to let that all settle before trying to consolidate everything into `TemplateActions`. With the long term view to make everything a "pattern" does it also make sense to rename `TemplateActions` to something like `PatternActions` as another step toward the great unification?

**Comment by ramonjd** [comment]
Thanks for the feedback @aaronrobertshaw 

> It makes sense to me to let that all settle before trying to consolidate everything into TemplateActions. 

Good to know thanks.

> With the long term view to make everything a "pattern" does it also make sense to rename TemplateActions to something like PatternActions as another step toward the great unification

Yes, I marked it as a follow up, but can add it to this PR give it's not a huge change.

Cheers.

**Comment by t-hamano** [comment]
I would like to inform you that #56392 requesting the functionality achieved by this PR has been submitted 👍

**Comment by richtabor** [comment]
Is there a reason to wait on this now?

**Comment by ramonjd** [comment]
> Is there a reason to wait on this now?

There could be some optimizations in this PR that may not apply after all the work being done in this area, or maybe more consolidation opportunities. 🤷🏻 

Checking with @glendaviesnz and @aaronrobertshaw 

I'll give it rebase sometime this week and retest.

File: packages/edit-site/src/components/page-patterns/grid-item.js [hunk 6]
```diff
@@ -86,49 +66,9 @@ function GridItem( { categoryId, item, ...props } ) {
...
-					? error.message
-					: __( 'An error occurred while deleting the pattern.' );
-			createErrorNotice( errorMessage, {
-				type: 'snackbar',
-				id: 'edit-site-patterns-error',
-			} );
-		}
-	};
-	const deleteItem = () =>
-		isTemplatePart ? removeTemplate( item ) : deletePattern();
```
Comment (ramonjd) [inline packages/edit-site/src/components/page-patterns/grid-item.js:112]: TODO:

Still working out whether we still need `removeTemplate` for clearing customizations to theme templates/patterns.

Currently, this PR uses `revertTemplate`. Does it have the same effect?

File: packages/edit-site/src/components/pattern-actions/index.js [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+	const { __experimentalDeleteReusableBlock } =
+		useDispatch( reusableBlocksStore );
+	// Only custom patterns or custom template parts can be renamed or deleted.
+	const isUserPattern = record?.type === PATTERN_TYPES.user;
+	const isTemplate = record?.type === TEMPLATE_POST_TYPE;
+	const isTemplatePart = record?.type === TEMPLATE_PART_POST_TYPE;
+
+	if ( ! isTemplatePart && ! isTemplate && ! isUserPattern ) {
+		return null;
+	}
```
Comment (t-hamano) [inline packages/edit-site/src/components/pattern-actions/index.js]: I noticed that the Duplicate menu was missing from the theme's pattern preview. This is because `isUserPattern` is `false`. Furthermore, `shouldDisplayMenu` after this will also be `false`, so the dropdown will not be rendered. If the theme has patterns, we may need to adjust the conditional statement to render only the Duplicate menu.

theme-pattern

Comment (ramonjd) [reply packages/edit-site/src/components/pattern-actions/index.js]: Good spotting! I'll update.

This PR is still a work in progress so I think there'll be a few things like that we'll have to tweak.

Comment (ramonjd) [reply packages/edit-site/src/components/pattern-actions/index.js]: Thanks for testing @t-hamano 

I think I've addressed this.

File: packages/edit-site/src/components/page-patterns/grid-item.js [hunk 8]
```diff
@@ -265,66 +196,8 @@ function GridItem( { categoryId, item, ...props } ) {
...
-								>
-									{ hasThemeFile
-										? __( 'Clear customizations' )
-										: __( 'Delete' ) }
-								</MenuItem>
-							) }
-						</MenuGroup>
-					) }
-				</DropdownMenu>
+				<PatternActions item={ item } />
```
Comment (t-hamano) [inline packages/edit-site/src/components/page-patterns/grid-item.js]:
```suggestion
				<PatternActions
					toggleProps={ { className: 'edit-site-patterns__button' } }
					item={ item }
				/>
```

It seems that the pattern action button is no longer visible on the pattern page. To apply [this style](https://github.com/WordPress/gutenberg/blob/2642697ff65f6f011044f8939f9f7ed70af2945c/packages/edit-site/src/components/page-patterns/style.scss#L166-L168), give the toggle button a class name.

pattern-dropdown-button

File: packages/edit-site/src/components/pattern-actions/index.js [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+		! isNonUserPattern
+	) {
+		return null;
+	}
+
+	const isRemovable = isTemplateRemovable( record );
+	const isRevertable = isTemplateRevertable( record );
+	const isEditable = isUserPattern || isRemovable;
+	// Only patterns and template parts can be duplicated for now.
+	const isDuplicable = isUserPattern || isNonUserPattern || isTemplatePart;
```
Comment (t-hamano) [inline packages/edit-site/src/components/pattern-actions/index.js:96]: With this condition, it seems that duplication of template parts of the theme is also allowed. Is this intentional?

duplicate-template-part

Furthermore, hybrid themes seem to allow duplication of theme template parts. You can check it by following the steps below.

- Access `http://localhost:8889/wp-admin` (instead of `http://localhost:8888/wp-admin`)
- Activate Emptyhybrid theme (this theme supports block-template-parts)
- Access Appearance > Template Parts
  - Expected (trunk): No action button is displayed.
  - Actual (this PR): Can duplicate template parts of the theme

File: packages/edit-site/src/components/pattern-actions/index.js [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+			);
+			const errorMessage =
+				error.message && error.code !== 'unknown_error'
+					? error.message
+					: fallbackErrorMessage;
+
+			createErrorNotice( errorMessage, { type: 'snackbar' } );
+		}
+	}
+
```
Comment (t-hamano) [inline packages/edit-site/src/components/pattern-actions/index.js:185]: It appears that an empty dropdown menu is rendered when there are no actions available.

no-action

We may be able to come up with a better conditional statement, but if there is nothing that can be rendered, it seems better to return null.

```javascript
const showDropdown =
	isEditable ||
	isDuplicable ||
	isUserPattern ||
	isEditable ||
	isRevertable;

if ( ! showDropdown ) {
	return null;
}
```

**Review by t-hamano — COMMENTED** [review]
I think this PR is still in progress, but I would like to check it out again and share what I noticed 👍

**Review by t-hamano — CHANGES_REQUESTED** [review]
Update as "Request changes" to prevent unintentional merging.

File: packages/edit-site/src/components/pattern-actions/index.js (hunk 9 above)
Comment (ramonjd) [reply packages/edit-site/src/components/pattern-actions/index.js:96]: > With this condition, it seems that duplication of template parts of the theme is also allowed. Is this intentional?

Yes. It looks like we can duplicate theme template parts on trunk as well so I wanted to harmonize the behaviour.

Screenshot 2023-12-11 at 6 56 57 pm

File: packages/edit-site/src/components/page-patterns/grid-item.js (hunk 8 above)
Comment (ramonjd) [reply packages/edit-site/src/components/page-patterns/grid-item.js]: Oh good spotting. Thanks 👍🏻

**Comment by ramonjd** [comment]
> I think this PR is still in progress, but I would like to check it out again and share what I noticed 👍

Yes, I've been neglecting this PR 😅 

Thanks a lot for your help in finding the bugs!! 🙇🏻

File: packages/edit-site/src/components/pattern-actions/index.js (hunk 9 above)
Comment (t-hamano) [reply packages/edit-site/src/components/pattern-actions/index.js:96]: > It looks like we can duplicate theme template parts on trunk as well so I wanted to harmonize the behaviour.

I had missed being able to do something like that 😅 Then I think this behavior is correct. However, I think we need to disable the duplicate action for classic themes that support block template parts.

Comment (ramonjd) [reply packages/edit-site/src/components/pattern-actions/index.js:96]: Good call. I've prevented non-block based themes from rendering the drop down. Is that the best way forward do you think?

File: packages/edit-site/src/components/editor/index.js [hunk 11]
```diff
@@ -27,6 +27,7 @@ import {
 	EditorKeyboardShortcuts,
 	EditorNotices,
 	EditorSnackbars,
+	PostLockedModal,
```
Comment (t-hamano) [inline packages/edit-site/src/components/editor/index.js]: Maybe this is an unintended change?

**Review by t-hamano — COMMENTED** [review]
From what I've tested, I think it's almost ready for shipping 👍

- Patterns Page:
  - Theme patterns and Theme template parts can only be duplicated
  - User patterns can be renamed, duplicated, exported, and deleted
  - Custom template part can berenamed, duplicated, and deleted
  - Page doesn't move anywhere after deleting pattern and template part
- Pattern Details Page Sidebar:
  - Can be renamed, duplicated, exported, and deleted
  - Return to the Patterns page after deleting the pattern
- Template Parts List Page:
  - Theme template parts can only be duplicated (This is intentional, as stated in [this comment](https://github.com/WordPress/gutenberg/pull/54271#discussion_r1422052851))
  - Can clear the theme's template part customization.
  - Custom template part can be renamed, duplicated, and deleted
  - Page doesn't move anywhere after deleting pattern
  - Classic themes that support template parts do not display the action button.
- Template Part Details Page Sidebar:
  - Custom template part can be renamed, duplicated, and deleted
  - 🤔 Return to the Patterns page after deleting the template part. In trunk, return to the Template Parts List page.
  - Classic themes that support template parts do not display the action button.
- Templates List Page:
  - Template pages that use DataView have been stabilized by #57109, but this PR should have no impact as this page uses [internal actions](https://github.com/WordPress/gutenberg/blob/05b7175a7b0d51ffe9c79f02b9839c4bddd7431c/packages/edit-site/src/components/page-templates/template-actions.js). As a follow-up, we may be able to standardize the actions on this page.
- Experimental Pages List Page:
  - This PR should have no impact as the action uses [a different component](https://github.com/WordPress/gutenberg/blob/025d764afdc6c1b093c410d528523d62f2a99b74/packages/edit-site/src/components/actions/index.js)
- Others:
  - The text field and button height of the Rename modal is 40px, so there are no regressions from the previous design improvements made in #56933.
  - All rename actions render correctly when using emojis in the name

File: packages/edit-site/src/components/page-patterns/grid-item.js (hunk 6 above)
Comment (t-hamano) [reply packages/edit-site/src/components/page-patterns/grid-item.js:112]: As far as I tested it, there was no problem, but I'm a little unsure. Does @aaronrobertshaw know more about this?

File: packages/edit-site/src/components/editor/index.js (hunk 11 above)
Comment (ramonjd) [reply packages/edit-site/src/components/editor/index.js]: Bad rebase. Thanks for spotting it!

**Comment by ramonjd** [comment]
Thank you for testing so thoroughly @t-hamano 🙇🏻

**Comment by ramonjd** [comment]
> 🤔 Return to the Patterns page after deleting the template part. In trunk, return to the Template Parts List page.

Just to clarify, in this PR, after you delete a template part, it'll take you back to the previous screen:

- after deleting directly from the Template Parts List page, you'll remain on the Template Parts List page
- clicking on a custom template on the Template Parts List page, and then deleting, you'll return to the Template Parts List page
- after deleting directly from the patterns page, you'll remain on the patterns page
- clicking on a custom template part on the patterns page, and then deleting you'll return to the patterns page

**Comment by t-hamano** [comment]
> > 🤔 Return to the Patterns page after deleting the template part. In trunk, return to the Template Parts List page.
> 
> Just to clarify, in this PR, after you delete a template part, it'll take you back to the previous screen:

Ah, I tried again and was unable to reproduce the issue. Maybe my testing procedure was wrong 🙈

**Comment by ramonjd** [comment]
> Ah, I tried again and was unable to reproduce the issue. Maybe my testing procedure was wrong

Ha, I don't trust myself. But what I think you saw is right? I tested again and it seems okay 🤷🏻 

I might get some folks who worked on patterns to also give it a test when they have free time.

**Comment by t-hamano** [comment]
>Ha, I don't trust myself. But what I think you saw is right? I tested again and it seems okay 🤷🏻

I tried it again and it seems to be fine.

https://github.com/WordPress/gutenberg/assets/54422211/1410cd22-9c94-4d23-bfa0-51465d9a8da2

File: packages/edit-site/src/components/page-patterns/grid-item.js (hunk 6 above)
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/page-patterns/grid-item.js:112]: My memory is a bit foggy on this one, so take this with a grain of salt.

As I understand it, there is a difference between `removeTemplate` and `revertTemplate` in that if the template doesn't have a theme file it won't be classed as revertable and an error notice is displayed with no further action taken.

If it's testing ok as Aki mentions then I'm probably missing something.

File: packages/edit-site/src/components/pattern-actions/index.js [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+	) {
+		return null;
+	}
+
+	const isRemovable = isTemplateRemovable( record );
+	const isRevertable = isTemplateRevertable( record );
+	// Only patterns and template parts can be duplicated for now.
+	const isDuplicable = isUserPattern || isNonUserPattern || isTemplatePart;
+
+	// If the pattern is not editable or duplicable, don't show the menu.
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/pattern-actions/index.js:98]:
```suggestion
	// If the pattern is not removable, revertable, or duplicable, don't show the menu.
```

The current comment doesn't quite seem to fit given `isEditable` is defined below and not used in the conditional.

Alternatively, is it worth doing away with the comment entirely as that code block is pretty self-explanatory now?

File: packages/edit-site/src/components/pattern-actions/index.js [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+							item={ record || item }
+							onClose={ onClose }
+							label={ __( 'Duplicate' ) }
+						/>
+					) }
+					{ isUserPattern && (
+						<MenuItem onClick={ () => exportAsJSON( record ) }>
+							{ __( 'Export as JSON' ) }
+						</MenuItem>
+					) }
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/pattern-actions/index.js:224]: If we're looking to increase parity, should we also be able to export/import user-created template parts?

File: packages/edit-site/src/components/pattern-actions/index.js [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+							} }
+						/>
+					) }
+				</MenuGroup>
+			) }
+		</DropdownMenu>
+	);
+}
+
+function DeleteMenuItem( { onRemove, title } ) {
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/pattern-actions/index.js:250]: Is it worth also splitting out the `DeleteMenuItem` and `RevertMenuItem` components to their own files to match `DuplicateMenuItem` & `RenameMenuItem`?

File: packages/edit-site/src/components/pattern-actions/rename-menu-item.js [hunk 15]
```diff
@@ -19,29 +19,48 @@ import { decodeEntities } from '@wordpress/html-entities';
...
+	TEMPLATE_PART_POST_TYPE,
+	PATTERN_TYPES,
+	TEMPLATE_ORIGINS,
+	POST_TYPE_LABELS,
+} from '../../utils/constants';
 
-export default function RenameMenuItem( { template, onClose } ) {
-	const title = decodeEntities( template.title.rendered );
-	const [ editedTitle, setEditedTitle ] = useState( title );
+export default function RenameMenuItem( { postType, postId, onClose } ) {
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/pattern-actions/rename-menu-item.js:30]: While we're trying to make things more consistent, I noticed this gets passed the type and id separately, but in other places, we're passing an item object. 

It might be good to tweak that so all these related components are consumed in an intuitive manner. Would it also prevent any excessive re-renders, not passing a fresh object around?

What do you think?

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for the continued effort here @ramonjd and the thorough testing @t-hamano 🙇 

After a quick test and look over the code, I've left a few random thoughts and questions. I'll have more time next week to dig in and test further.

For the most part though this seems to be working pretty well. One thing that stuck out to me was the new inclusion of help text into the actions menu for clearing customizations. 

It looks a bit cluttered to me and could have been the reason it was omitted when added for the patterns page in trunk. I appreciate that it is present in the template part list view's action menu, so perhaps we need some design feedback on this one?

| Trunk | This PR |
|---|---|
| Screenshot 2023-12-22 at 5 40 24 pm | Screenshot 2023-12-22 at 5 39 35 pm |


> renames 'Template Actions' to 'Pattern actions' in anticipation of the [great unification](https://github.com/WordPress/gutenberg/pull/54271#issuecomment-1760612399).

P.S. We might need a new moniker for uniting all the different types of patterns as the unification of the site and post editor has sort of usurped the name. See https://github.com/WordPress/gutenberg/pull/57214 🙂

File: packages/edit-site/src/components/pattern-actions/rename-menu-item.js (hunk 15 above)
Comment (ramonjd) [reply packages/edit-site/src/components/pattern-actions/rename-menu-item.js:30]: This makes sense. Thanks for flagging. I'll take a look.

File: packages/edit-site/src/components/pattern-actions/index.js (hunk 13 above)
Comment (ramonjd) [reply packages/edit-site/src/components/pattern-actions/index.js:224]: Sounds good to me. It's not a big change for this PR.

File: packages/edit-site/src/components/pattern-actions/index.js (hunk 12 above)
Comment (ramonjd) [reply packages/edit-site/src/components/pattern-actions/index.js:98]: I was using a editable to cover all the change of state conditions, but you're right: it's just confusing noise since the vars are self-explanatory. I think deleting is a good option. 

👍🏻

**Comment by ramonjd** [comment]
Thanks again for looking at this @aaronrobertshaw and @t-hamano 

I'll tweak the PR as suggested. 

One parallel effort that may put the brakes on this PR further is:

- https://github.com/WordPress/gutenberg/pull/57333#issuecomment-1867542519

The data view components take a different approach to rendering action menu. 

I'm not 100% sure yet, but they may not be compatible with existing action components such as these and the ones they replace. See: https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/page-templates/index.js#L335

So if the objective is to, similar to the manage all template page, refactor patterns and other views to use dataviews, then maybe this PR isn't worth pursuing in its current form.

**Comment by aaronrobertshaw** [comment]
> So if the objective is to, similar to the manage all template page, refactor patterns and other views to use dataviews, then maybe this PR isn't worth pursuing in its current form.

Appreciate the effort you've put into this one. I have no objections if you want to close this PR, it can always be resurrected if needed.

**Comment by ramonjd** [comment]
I'll close this PR - as implied [above](https://github.com/WordPress/gutenberg/pull/54271#issuecomment-1869753850), it might create more refactor headaches as we migrate to data views.

Also, the PR has been open for a while and it probably too large as it is. Therefore, I don't see the continued effort as providing commensurate value. 

I still think it's desirable to harmonize the actions between screens, but perhaps in a less ambitious way, and with consideration of the way dataviews consume actions.

Thanks again for all the help along the way folks!
